File: Blockbuster-parent/Blockbuster-model/src/main/java/net/dualsoft/blockbuster/model/service/AuthService.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package net.dualsoft.blockbuster.model.service;

import net.dualsoft.blockbuster.db.jooq.tables.pojos.Customer;
import net.dualsoft.blockbuster.model.helper.pojos.Response;
import net.dualsoft.blockbuster.db.jooq.tables.pojos.Staff;
import net.dualsoft.blockbuster.model.helper.pojos.CustomerReg;

/**
 *
 * @author teodora
 */
public interface AuthService {
    
    public Response<Customer> login(String email, String password);
    
    public Response<Customer> register(CustomerReg customer);
    
    public Response<Staff> loginStaff(String email, String password);
    
    public Response changePassword(String code, String newPassword);
    
    public Response editPassword(int customerId, String oldPassword, String newPassword);
    
    public Response getLockedFunds(int customerId);
}

File: Blockbuster-parent/blockbuster-mobile/src/app/_services/auth.service.ts
import { tap } from 'rxjs/operators';
import { Customer } from './../_models/customer.model';
import { HttpClient, HttpHeaders, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable, Subject } from 'rxjs';
import { Auth } from '../_models/auth.model';
import { Response } from '../_models/http/response.model';
import { DomainService } from './domain/domain.service';

@Injectable({
  providedIn: 'root',
})
export class AuthService {
  private authSubject: Subject<Auth> = new Subject<Auth>();

  url = this.domain.getAddress() + '/auth';

  constructor(private http: HttpClient, private domain: DomainService) {}

  login(email: string, password: string): Observable<Response<Auth>> {
    return this.http.post<Response<Auth>>(
      `${this.url}/login`,
      { email, password },
      {
        withCredentials: true,
        headers: new HttpHeaders({ 'Content-Type': 'application/json' }),
      }
    );
  }

  publishAuth(auth: Auth) {
    this.authSubject.next(auth);
  }

  getAuthObservable() {
    return this.authSubject.asObservable();
  }

  register(customer: Customer): Observable<Response<Auth>> {
    return this.http
      .post<Response<Auth>>(`${this.url}/register`, customer, {
        withCredentials: true,
        headers: new HttpHeaders({ 'Content-Type': 'application/json' }),
      })
      .pipe(tap(console.log));
  }

  requestPasswordReset(email: string) {
    return this.http.post(
      `${this.domain.getAddress()}/mail/changePassword`,
      email
    );
  }

  passwordReset(code: string, password: string) {
    return this.http.post(`${this.url}/changePassword`, { password, code });
  }

  editPassword(customerId: number, oldPassword: string, newPassword: string) {
    return this.http.post(`${this.url}/editPassword`, {
      customerId,
      oldPassword,
      newPassword,
    });
  }

  getLockedFunds(customerId: number) {
    return this.http.get<Response<number>>(
      `${this.url}/lockedFunds/${customerId}`
    );
  }
}
